<a href="https://colab.research.google.com/github/RhythmSanghi23/Data-Analysis-and-Visualization/blob/main/Data_Analysis_and_Visualization__FinalAssignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Task 1: Data Loading and Data Aggregation**






*   Load the 3 data files into the variables data_18, data_19, data_20.

In [186]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
data_18 = pd.read_csv("https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2018.csv ")
data_19 = pd.read_csv("https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2019.csv ")
data_20 = pd.read_csv("https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2020.csv ")

In [187]:
data_18

Timestamp   Age  ... Company size   Company type
0    14/12/2018 12:41:33  43.0  ...     100-1000        Product
1    14/12/2018 12:42:09  33.0  ...       50-100        Product
2    14/12/2018 12:47:36  32.0  ...        1000+        Product
3    14/12/2018 12:50:15  25.0  ...        1000+        Product
4    14/12/2018 12:50:31  39.0  ...     100-1000  Ecom retailer
..                   ...   ...  ...          ...            ...
760  03/06/2020 20:12:51  40.0  ...        10-50        Product
761  28/07/2020 04:03:13   NaN  ...        10-50        Product
762  28/07/2020 04:03:20   NaN  ...        10-50        Product
763  26/08/2020 09:06:44   NaN  ...          NaN            NaN
764  21/09/2020 01:47:10  31.0  ...        1000+        Product

[765 rows x 14 columns]


*   Data aggregation is the process of gathering data and presenting it in a summarized format. The data may be gathered from multiple data sources with the intent of combining these data sources into a summary for data analysis.
Similar to how this dataset involves 3 data files, you’ll often be working on combining information from 2 or more files and analysing it. More often than not, GroupBy is a very useful tool for this purpose.

In [188]:
data18_19=pd.concat([data_18,data_19],axis=1)
result=pd.concat([data_20,data18_19],axis=1)
result

Timestamp   Age  ... Company business sector   0
0     24/11/2020 11:14:15  26.0  ...                 Tourism NaN
1     24/11/2020 11:14:16  26.0  ...   Scientific Activities NaN
2     24/11/2020 11:14:21  29.0  ...                     NaN NaN
3     24/11/2020 11:15:24  28.0  ...               Transport NaN
4     24/11/2020 11:15:46  37.0  ...              Automotive NaN
...                   ...   ...  ...                     ...  ..
1248  18/01/2021 15:05:50  31.0  ...                     NaN NaN
1249  18/01/2021 17:46:02  33.0  ...                     NaN NaN
1250  18/01/2021 23:20:35  39.0  ...                     NaN NaN
1251  19/01/2021 10:17:58  26.0  ...                     NaN NaN
1252  19/01/2021 12:01:11  26.0  ...                     NaN NaN

[1253 rows x 60 columns]

In [189]:
data_18.describe()


Age  ...  Salary two years ago
count  672.000000  ...            463.000000
mean    32.183036  ...          58013.475162
std      5.107268  ...          20413.048908
min     21.000000  ...          10001.000000
25%     29.000000  ...          48000.000000
50%     32.000000  ...          56000.000000
75%     35.000000  ...          67000.000000
max     60.000000  ...         150000.000000

[8 rows x 5 columns]

# **Task 2: Data Analysis**





* Display the first 5 rows of the 2018 survey data

In [190]:
data_18.head(5)

Timestamp   Age  ... Company size   Company type
0  14/12/2018 12:41:33  43.0  ...     100-1000        Product
1  14/12/2018 12:42:09  33.0  ...       50-100        Product
2  14/12/2018 12:47:36  32.0  ...        1000+        Product
3  14/12/2018 12:50:15  25.0  ...        1000+        Product
4  14/12/2018 12:50:31  39.0  ...     100-1000  Ecom retailer

[5 rows x 14 columns]


*   Display a concise summary of the 2020 data and list out 3 observations/inferences that you observe from the result. For this you will need to use the info() method.



In [191]:
data_20.info()
data_20.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1253 entries, 0 to 1252
Data columns (total 23 columns):
 #   Column                                                                                                                   Non-Null Count  Dtype  
---  ------                                                                                                                   --------------  -----  
 0   Timestamp                                                                                                                1253 non-null   object 
 1   Age                                                                                                                      1226 non-null   float64
 2   Gender                                                                                                                   1243 non-null   object 
 3   City                                                                                                                     1253 non-null   o

Timestamp  ...  Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR
0  24/11/2020 11:14:15  ...                                                NaN                                                                      
1  24/11/2020 11:14:16  ...                                                NaN                                                                      
2  24/11/2020 11:14:21  ...                                                NaN                                                                      
3  24/11/2020 11:15:24  ...                                                NaN                                                                      
4  24/11/2020 11:15:46  ...                                                NaN                                                                      

[5 rows x 23 columns]

In [192]:
data_20.describe()

Age  ...  Have you been forced to have a shorter working week (Kurzarbeit)? If yes, how many hours per week
count  1226.000000  ...                                         373.000000                                                
mean     32.509788  ...                                          12.967828                                                
std       5.663804  ...                                          15.275174                                                
min      20.000000  ...                                           0.000000                                                
25%      29.000000  ...                                           0.000000                                                
50%      32.000000  ...                                           0.000000                                                
75%      35.000000  ...                                          30.000000                                                
max      69.000000  ...                                          40.000000                                                

[8 rows x 4 columns]

In [193]:
data_20.describe(include=['object'])

Timestamp  ... Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR
count                  1253  ...                                                462                                                                     
unique                 1248  ...                                                 59                                                                     
top     24/11/2020 14:07:23  ...                                                  0                                                                     
freq                      2  ...                                                161                                                                     

[4 rows x 19 columns]

In [194]:
data_20.describe(include='all')

Timestamp  ...  Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR
count                  1253  ...                                                462                                                                      
unique                 1248  ...                                                 59                                                                      
top     24/11/2020 14:07:23  ...                                                  0                                                                      
freq                      2  ...                                                161                                                                      
mean                    NaN  ...                                                NaN                                                                      
std                     NaN  ...                                                NaN                                                                      
min                     NaN  ...                                                NaN                                                                      
25%                     NaN  ...                                                NaN                                                                      
50%                     NaN  ...                                                NaN                                                                      
75%                     NaN  ...                                                NaN                                                                      
max                     NaN  ...                                                NaN                                                                      

[11 rows x 23 columns]

This data is an anonymous salary survey that has been conducted for the year 2020 among European IT specialists with a stronger focus on Germany. This year 1238 respondents volunteered to participate in the survey.
The three observations seen so far from the data is that:
* The data has 1253 entries with 23 coloumns
* The minimum age given in the data is 20 and the maximum age is 69.
* The top position in the data is software engineer with the frequency of 387 out of 1253.





*   Display the descriptive statistics of the 2018 survey data



In [195]:
data_18.info()
data_18.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 765 entries, 0 to 764
Data columns (total 14 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Timestamp                           765 non-null    object 
 1   Age                                 672 non-null    float64
 2   Gender                              751 non-null    object 
 3   City                                736 non-null    object 
 4   Position                            737 non-null    object 
 5   Years of experience                 732 non-null    float64
 6   Your level                          743 non-null    object 
 7   Current Salary                      750 non-null    float64
 8   Salary one year ago                 596 non-null    float64
 9   Salary two years ago                463 non-null    float64
 10  Are you getting any Stock Options?  742 non-null    object 
 11  Main language at work               750 non-n

Age  ...  Salary two years ago
count  672.000000  ...            463.000000
mean    32.183036  ...          58013.475162
std      5.107268  ...          20413.048908
min     21.000000  ...          10001.000000
25%     29.000000  ...          48000.000000
50%     32.000000  ...          56000.000000
75%     35.000000  ...          67000.000000
max     60.000000  ...         150000.000000

[8 rows x 5 columns]

* Display the number of missing values in each column of the 2018 survey data How many people responded to the survey in each of the 3 years? Has the number increased or decreased over the years?

In [196]:
data_18.dtypes=='object'

Timestamp                              True
Age                                   False
Gender                                 True
City                                   True
Position                               True
Years of experience                   False
Your level                             True
Current Salary                        False
Salary one year ago                   False
Salary two years ago                  False
Are you getting any Stock Options?     True
Main language at work                  True
Company size                           True
Company type                           True
dtype: bool

In [197]:
num_vars= data_18.columns[data_18.dtypes!='object']
cat_vars=data_18.columns[data_18.dtypes=='object']

data_18[num_vars].isnull().sum()
data_18[cat_vars].isna().sum()

Timestamp                              0
Gender                                14
City                                  29
Position                              28
Your level                            22
Are you getting any Stock Options?    23
Main language at work                 15
Company size                          15
Company type                          35
dtype: int64

In [198]:
len(data_18),len(data_19),len(data_20)

(765, 991, 1253)

As shown above the number of people who responded in the survey have increased.

* Display all the unique values and their frequency in the column - “Number of vacation days” of 2020 data. Write down your observations (at least one) for this result.

In [199]:
data_20['Number of vacation days'].describe()

count     1185
unique      45
top         30
freq       488
Name: Number of vacation days, dtype: object

Observation: The top number of vacations given are 30 days 

# **Task 3: Data Cleaning**

* Rename the column ‘Position ‘ in the 2020 data to ‘Position’. (without the blank space)

In [200]:
data_20.rename(columns = {'Position ':'Position'}, inplace = True)
data_20

Timestamp  ...  Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR
0     24/11/2020 11:14:15  ...                                                NaN                                                                      
1     24/11/2020 11:14:16  ...                                                NaN                                                                      
2     24/11/2020 11:14:21  ...                                                NaN                                                                      
3     24/11/2020 11:15:24  ...                                                NaN                                                                      
4     24/11/2020 11:15:46  ...                                                NaN                                                                      
...                   ...  ...                                                ...                                                                      
1248  18/01/2021 15:05:50  ...                                                NaN                                                                      
1249  18/01/2021 17:46:02  ...                                                  0                                                                      
1250  18/01/2021 23:20:35  ...                                                NaN                                                                      
1251  19/01/2021 10:17:58  ...                                                  0                                                                      
1252  19/01/2021 12:01:11  ...                                                600                                                                      

[1253 rows x 23 columns]

* Check for missing values in 2020 data for all the columns. If there are no missing values, proceed to the next step. If there are missing values in the dataset,
 
1. For categorical variables, fill the
missing values with the mode of the data. Remember if the data type of any variable is ‘object’, it is categorical variable.
2.For numerical variables, fill the missing values with the mean of the data.






In [201]:
#data_20.dtypes=='object'
num_vars20= data_20.columns[data_20.dtypes!='object']
cat_vars20=data_20.columns[data_20.dtypes=='object']

data_20[num_vars20].isnull().sum()
data_20[cat_vars20].isna().sum()

Timestamp                                                                                                                    0
Gender                                                                                                                      10
City                                                                                                                         0
Position                                                                                                                     6
Total years of experience                                                                                                   16
Years of experience in Germany                                                                                              32
Seniority level                                                                                                             12
Your main technology / programming language                                                                    

In [202]:
mean=data_20['Age'].mean()
mean


32.50978792822186

In [203]:
data_20[num_vars20].dropna(axis='columns', how='all')
data_20[num_vars20].fillna(data_20.mean(axis='columns'))


Age  ...  Have you been forced to have a shorter working week (Kurzarbeit)? If yes, how many hours per week
0     26.0  ...                                                NaN                                                
1     26.0  ...                                                NaN                                                
2     29.0  ...                                                NaN                                                
3     28.0  ...                                                NaN                                                
4     37.0  ...                                                NaN                                                
...    ...  ...                                                ...                                                
1248  31.0  ...                                                NaN                                                
1249  33.0  ...                                                NaN                                                
1250  39.0  ...                                                NaN                                                
1251  26.0  ...                                                NaN                                                
1252  26.0  ...                                               30.0                                                

[1253 rows x 4 columns]


*   Drop the timestamp column for all the three years data since the date and time at which a person filled the survey is irrelevant to us. The year matters and we already know that from the dataset’s name.





In [204]:
data_20[cat_vars20].apply(lambda x: x.fillna(x.value_counts().index[0]))

Timestamp  ... Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR
0     24/11/2020 11:14:15  ...                                                  0                                                                     
1     24/11/2020 11:14:16  ...                                                  0                                                                     
2     24/11/2020 11:14:21  ...                                                  0                                                                     
3     24/11/2020 11:15:24  ...                                                  0                                                                     
4     24/11/2020 11:15:46  ...                                                  0                                                                     
...                   ...  ...                                                ...                                                                     
1248  18/01/2021 15:05:50  ...                                                  0                                                                     
1249  18/01/2021 17:46:02  ...                                                  0                                                                     
1250  18/01/2021 23:20:35  ...                                                  0                                                                     
1251  19/01/2021 10:17:58  ...                                                  0                                                                     
1252  19/01/2021 12:01:11  ...                                                600                                                                     

[1253 rows x 19 columns]

In [205]:
num_vars19= data_19.columns[data_19.dtypes!='object']
cat_vars19=data_19.columns[data_19.dtypes=='object']
data_19[num_vars19].dropna(axis='columns', how='all')
data_19[num_vars19].fillna(data_19.mean(axis='columns'))
data_19[cat_vars19].apply(lambda x: x.fillna(x.value_counts().index[0]))

Zeitstempel Gender  ... Сontract duration Company business sector
0    02.12.2019 11:18:26   Male  ...         unlimited                 Tourism
1    02.12.2019 11:18:35   Male  ...         unlimited   Scientific Activities
2    02.12.2019 11:18:56   Male  ...         unlimited                Сommerce
3    02.12.2019 11:19:08   Male  ...         unlimited               Transport
4    02.12.2019 11:19:37   Male  ...         unlimited              Automotive
..                   ...    ...  ...               ...                     ...
986  07.01.2020 09:23:01   Male  ...         unlimited                 Telecom
987  07.01.2020 10:08:18   Male  ...         unlimited     Finance / Insurance
988  07.01.2020 16:52:43   Male  ...         unlimited                Сommerce
989  08.01.2020 11:18:41   Male  ...  more than 1 year                  Health
990  08.01.2020 22:20:08   Male  ...         unlimited               Transport

[991 rows x 12 columns]

In [206]:

data_18[num_vars].dropna(axis='columns', how='all')
data_18[num_vars].fillna(data_18.mean(axis='columns'))
data_18[cat_vars].apply(lambda x: x.fillna(x.value_counts().index[0]))

Timestamp Gender  ... Company size   Company type
0    14/12/2018 12:41:33      M  ...     100-1000        Product
1    14/12/2018 12:42:09      F  ...       50-100        Product
2    14/12/2018 12:47:36      M  ...        1000+        Product
3    14/12/2018 12:50:15      M  ...        1000+        Product
4    14/12/2018 12:50:31      M  ...     100-1000  Ecom retailer
..                   ...    ...  ...          ...            ...
760  03/06/2020 20:12:51      M  ...        10-50        Product
761  28/07/2020 04:03:13      M  ...        10-50        Product
762  28/07/2020 04:03:20      M  ...        10-50        Product
763  26/08/2020 09:06:44      M  ...     100-1000        Product
764  21/09/2020 01:47:10      F  ...        1000+        Product

[765 rows x 9 columns]

In [207]:
data_20.drop(['Timestamp'], axis = 1)
data_19.drop(['Zeitstempel'], axis = 1)
data_18.drop(['Timestamp'], axis = 1)

Age Gender     City  ... Main language at work  Company size   Company type
0    43.0      M  München  ...               Deutsch      100-1000        Product
1    33.0      F  München  ...               Deutsch        50-100        Product
2    32.0      M  München  ...               Deutsch         1000+        Product
3    25.0      M  München  ...               English         1000+        Product
4    39.0      M  München  ...               English      100-1000  Ecom retailer
..    ...    ...      ...  ...                   ...           ...            ...
760  40.0      M     Köln  ...               Deutsch         10-50        Product
761   NaN      M     Köln  ...               Deutsch         10-50        Product
762   NaN      M     Köln  ...               Deutsch         10-50        Product
763   NaN    NaN      NaN  ...                   NaN           NaN            NaN
764  31.0      F  München  ...               English         1000+        Product

[765 rows x 13 columns]

# **Task 4: Data Visualization using Plotly**

* Create a pie chart to analyze the Company types in the year 2019. Are Consulting / Agency companies more popular than Startups? 

In [208]:
data_19

Zeitstempel   Age  ... Company business sector   0
0    02.12.2019 11:18:26  33.0  ...                 Tourism NaN
1    02.12.2019 11:18:35  29.0  ...   Scientific Activities NaN
2    02.12.2019 11:18:56   NaN  ...                     NaN NaN
3    02.12.2019 11:19:08  30.0  ...               Transport NaN
4    02.12.2019 11:19:37  32.0  ...              Automotive NaN
..                   ...   ...  ...                     ...  ..
986  07.01.2020 09:23:01  30.0  ...                 Telecom NaN
987  07.01.2020 10:08:18  28.0  ...     Finance / Insurance NaN
988  07.01.2020 16:52:43  42.0  ...                Сommerce NaN
989  08.01.2020 11:18:41  33.0  ...                  Health NaN
990  08.01.2020 22:20:08  32.0  ...               Transport NaN

[991 rows x 23 columns]

In [225]:
my_labels=data_19['Company type'].unique()
my_labels
my_data=data_19['Company type'].value_counts()
my_data
#my_data1 = [my_data[0],my_data[1],my_data[2],my_data[3],my_data[4],my_data[5],my_data[6]]
#my_data1
#plt.pie(my_data1,labels=my_labels,autopct='%1.1f%%')

Product                 619
Startup                 181
Consulting / Agency     117
Bodyshop / Outsource     30
University                6
Bank                      6
Outsource                 1
Name: Company type, dtype: int64

In [226]:
from collections import Counter
print(Counter(data_19['Company type']))
plt.figure(1, figsize=(20,10)) 
the_grid = GridSpec(2, 2)

Counter({'Product': 619, 'Startup': 181, 'Consulting / Agency': 117, nan: 31, 'Bodyshop / Outsource': 30, 'University': 6, 'Bank': 6, 'Outsource': 1})
